lib/matplotlib/tests/test_collections.py

```diff
@@ -1001,30 +1001,30 @@ def test_color_logic(pcfunc):
     # Reset edgecolor to default.
     pc.set_edgecolor(None)
     pc.update_scalarmappable()
-    assert mcolors.same_color(pc.get_edgecolor(), mapped)
+    assert np.array_equal(pc.get_edgecolor(), mapped)
     pc.set_facecolor(None)  # restore default for facecolor
     pc.update_scalarmappable()
-    assert mcolors.same_color(pc.get_facecolor(), mapped)
+    assert np.array_equal(pc.get_facecolor(), mapped)
     assert mcolors.same_color(pc.get_edgecolor(), 'none')
     # Turn off colormapping entirely:
     pc.set_array(None)
     pc.update_scalarmappable()
     assert mcolors.same_color(pc.get_edgecolor(), 'none')
     assert mcolors.same_color(pc.get_facecolor(), face_default)  # not mapped
     # Turn it back on by restoring the array (must be 1D!):
-    pc.set_array(z.ravel())
+    pc.set_array(z.ravel() if pcfunc == plt.pcolor else z)
```

Comment:
Both of these situations feel undesirable...
* The `must be 1D!` is the way it currently is and doesn't feel right after we've initialized our array with the 2D version `pcolor(z)`.
* It also doesn't feel great to change this and have to do the `if pcfunc ...` and change between 1D and 2D depending on which pcolor incantation I chose.

Reply:
The flattening in pcolormesh is a historical artifact; I think the move to accept it as fundamentally 2-D is good.
Pcolor is rarely a better choice than pcolormesh.  When it is, it is likely because it can generate a set of
colored quadrilaterals that don't have to fill out a 2-D domain.  I think we should embrace this difference between
pcolormesh and pcolor, so I don't have a problem with your special-casing in the test.  I'm open to discussing
ideas for backwards-compatible improvements to pcolor, so that it could also take a 2-D array in set_array,
but it doesn't seem like a priority.